staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -489,9 +495,13 @@ func (cnc *CloudNodeController) syncNode(ctx context.Context, nodeName string) e
 			return err
 		}
 
+		removeCloudProviderTaintDelay.Observe(time.Since(newNode.ObjectMeta.CreationTimestamp.Time).Seconds())
+
 		// After adding, call UpdateNodeAddress to set the CloudProvider provided IPAddresses
 		// So that users do not see any significant delay in IP addresses being filled into the node
-		cnc.updateNodeAddress(ctx, newNode, instanceMetadata)
+		if cnc.updateNodeAddress(ctx, newNode, instanceMetadata) {
+			initializeNodeAddressDelay.Observe(time.Since(newNode.ObjectMeta.CreationTimestamp.Time).Seconds())
```

Comment:
/lgtm cancel

hold on, this is only executed once the node is initialized but addresses may not be added here and be added on the periodic loop

```
	// The periodic loop for updateNodeStatus polls the Cloud Provider periodically
	// to reconcile the nodes addresses and labels.
	go wait.Until(func() {
		if err := cnc.UpdateNodeStatus(context.TODO()); err != nil {
			klog.Errorf("failed to update node status: %v", err)
		}
	}, cnc.nodeStatusUpdateFrequency, stopCh)
```

also addresses are independent of the node lifecycle, they can be added at any time, that is why I suggested to add the metrics on the `updateNodeAddress` method, but I know see that it can be misleading

@logicalhan @dgrisonnet do you have any suggestion how to measure this?

Reply:
> this is only executed once the node is initialized but addresses may not be added here and be added on the periodic loop

Correct. Until `syncNode` executes, the `Node` has no addresses at all. This metric is measuring the amount of time that a `Node` has no addresses.

Measuring the periodic reconciliation of the addresses doesn't seem very helpful; we can't measure the amount of time it took the controller to update the addresses after they changed in the cloud provider's API, and [we decided in earlier revisions of this PR](https://github.com/kubernetes/kubernetes/pull/119241#discussion_r1261758474) that a simple timing of the update logic (which is just a cloud provider call and a write to the API server) isn't very informative.

Reply:
I would agree that it seems better to just measure the initial addressing delay from the cloud controller perspective.
However, from @aojea's explaination, it seems that it would be better to move the observation of the metric in `updateNodeAddress` and create a condition to check whether this is the first addressing or not.

Reply:
It may be tricky/messy to check which occurrence is the first node status update.  Would it be simpler to just have a gauge for when there's a taint removal or node address update event? It should not be super difficult to then compare this with the node creation time.

Reply:
(if we can have updateNodeAddress to detect the first update, that's fine too)

Reply:
> it seems that it would be better to move the observation of the metric in updateNodeAddress and create a condition to check whether this is the first addressing or not.

I don't agree, I think code like this:
```
// true when initial node sync
cnc.updateNodeAddress(ctx, newNode, instanceMetadata, true)
// false when periodic update
cnc.updateNodeAddress(ctx, newNode, instanceMetadata, false)
```

is less clear than:
```
if cnc.updateNodeAddress(ctx, newNode, instanceMetadata) {
    initializeNodeAddressDelay.Observe(time.Since(newNode.ObjectMeta.CreationTimestamp.Time).Seconds())
}
```

It's not more correct to spread the metric logic across 2 functions; and I think it's preferable to instead gather it in a single function.

Reply:
Having only:
```
if cnc.updateNodeAddress(ctx, newNode, instanceMetadata) {
    initializeNodeAddressDelay.Observe(time.Since(newNode.ObjectMeta.CreationTimestamp.Time).Seconds())
}
```
will not be consistent across all the nodes no? since some of them will have their addresses initiliazed by the periodic update

Reply:
> Would it be simpler to just have a gauge for when there's a taint removal or node address update event? It should not be super difficult to then compare this with the node creation time.

i don't think this would be simpler, no. You'd have to query both the API server (assuming the node still exists -- which is may not) *and* your metric backend to make that kind of comparison. You could accomplish the same thing by adding some log statements and grepping for them.

We're not really interested in how long a specific node took to sync; we're interested in computing a p50/p95/etc, and using that value to decide whether the configured concurrency of the node controller is sufficient (or not).

Reply:
I've removed this metric in the latest revision.

Reply:
sounds good to me :+1: